src/linpred.jl

```diff
@@ -219,8 +281,68 @@ function invchol(x::DensePredChol{T,<: CholeskyPivoted}) where T
     ipiv = invperm(ch.p)
     res[ipiv, ipiv]
 end
+
 invchol(x::SparsePredChol) = cholesky!(x) \ Matrix{Float64}(I, size(x.X, 2), size(x.X, 2))
-vcov(x::LinPredModel) = rmul!(invchol(x.pp), dispersion(x, true))
+
+function vcov(x::LinPredModel)
+    d = dispersion(x, true)
+    u = residuals(x; weighted = isweighted(x))
+    _vcov(x.pp, u, d)
+end
+
+_vcov(pp::LinPred, u, d) = rmul!(invchol(pp), d)
+
+function _vcov(pp::DensePredChol{T, <:Cholesky, <:ProbabilityWeights}, u::AbstractVector, d::Real) where {T}
+    wts = pp.wts
+    Z = mul!(pp.scratchm1, Diagonal(sqrt.(wts).*u), pp.X)
+    XtW2X = Z'Z
+    invXtWX = invchol(pp)
+    V = invXtWX*XtW2X*invXtWX
+    n = length(wts)
+    k = length(pp.delbeta)
+    n/(n-k)*V
+end
+
+function nancolidx(A::AbstractMatrix)
+    ## Return to set the idx:
+    ## idx_nancol idx_nonnancol
+    nrow, ncol = size(A)
+    idx_nancol = Int64[]
+    idx_nonnancol = Int64[]
+    for j in axes(A, 2)
+        h = 0
+        for i in axes(A, 1)
+            h += isnan(A[i,j]) ? 1 : 0
+        end
+        h == nrow ? push!(idx_nancol, j) : push!(idx_nonnancol, j)
+    end
+    return (idx_nancol, idx_nonnancol)
+end
+
+function _vcov(pp::DensePredChol{T, <:CholeskyPivoted, <:ProbabilityWeights}, u::AbstractVector, d::Real) where {T}
+    wts = pp.wts
+    Z = mul!(pp.scratchm1, Diagonal(sqrt.(wts).*u), pp.X)
+    ch = pp.chol
+    rnk = rank(ch)
+    p = length(pp.delbeta)
+    invXtWX = invchol(pp)
+    if rnk == p
+        B = Z'Z
+        A = invXtWX
+        V = A*B*A
+    else
+        idx_nan, idx_non = nancolidx(invXtWX) 
+        Zc = view(Z, :, idx_non)
+        B = Zc'Zc
+        A = view(invXtWX, idx_non, idx_non)
+        V = similar(pp.scratchm2)
+        V[idx_non, idx_non] = A*B*A
+        V[idx_nan, :] .= convert(T, NaN)
+        V[:, idx_nan] .= convert(T, NaN)
```

Comment:
I don't really get why all of this is needed. Could you explain why matrices do not have matching structures? Should we change the way they are stored somewhere else?

Reply:
Bump.